packages/rest-crud/src/crud-rest-builder.plugin.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/rest-crud
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {bind, ControllerClass, inject} from '@loopback/core';
+import {
+  asModelApiBuilder,
+  ModelApiBuilder,
+  ModelApiConfig,
+} from '@loopback/model-api-builder';
+import {
+  ApplicationWithRepositories,
+  Class,
+  Entity,
+  EntityCrudRepository,
+} from '@loopback/repository';
+import {Model} from '@loopback/rest';
+import debugFactory from 'debug';
+import {defineCrudRepositoryClass, defineCrudRestController} from '.';
+
+const debug = debugFactory('loopback:boot:crud-rest');
+
+export interface ModelCrudRestApiConfig extends ModelApiConfig {
+  // E.g. '/products'
+  basePath: string;
+}
+
+@bind(asModelApiBuilder)
+export class CrudRestApiBuilder implements ModelApiBuilder {
+  readonly pattern: string = 'CrudRest';
+
+  build(
+    application: ApplicationWithRepositories,
+    modelClass: typeof Model & {prototype: Model},
+    cfg: ModelApiConfig,
+  ): Promise<void> {
+    const modelName = modelClass.name;
+    const config = cfg as ModelCrudRestApiConfig;
+    if (!config.basePath) {
+      throw new Error(
+        `Missing required field "basePath" in configuration for model ${modelName}.`,
+      );
+    }
+
+    if (!(modelClass.prototype instanceof Entity)) {
+      throw new Error(
+        `CrudRestController requires an Entity, Models are not supported. (Model name: ${modelName})`,
```

Comment:
> `CrudRestController requires an Entity, Models are not supported. (Model name: ${modelName})`,

What are we trying to say here? That the model needs to extend Entity?

Perhaps

`CrudRestController requires a model that extends 'Entity'. (Model name: ${modelName} does not extend 'Entity')`

Reply:
Why are we calling the file 'plugin'?

> packages/rest-crud/src/crud-rest-builder.plugin.ts

Reply:
> Why are we calling the file 'plugin'?
> 
> > packages/rest-crud/src/crud-rest-builder.plugin.ts

Following the terminology in the [spike](https://github.com/strongloop/loopback-next/pull/3617/files#diff-1ac72022ee3c92074711610d69356271R49-R59)

Reply:
@nabdelgadir ,

Hmm. Let's not use the word plugin. We don't mention it in the overall LB4 docs.

As an example,  AuthenticationStrategyProvider is a class that defines an extension point where developers can create extensions (custom authentication strategies) and register them against the extension point... but we don't call any of the files  `.plugin.ts` . 

Same for the extensions for greeter extension point : https://github.com/strongloop/loopback-next/blob/master/examples/greeter-extension/src/greeters/greeter-en.ts#L6  , we don't use `.plugin.ts` .

https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/providers/auth-strategy.provider.ts#L28

Reply:
> Let's not use the word plugin. We don't mention it in the overall LB4 docs.

Considering that we are using ExtensionPoint/Extension design pattern, maybe we can use "extension" instead of "plugin"?

Reply:
On the other hand, since this file contains `CrudRestApiBuilder` class which is bound `asModelApiBuilder`, maybe `crud-rest.model-api-builder.ts` would be a good name, using `model-api-builder` as the type suffix (similar to `controller` and `repository` we use elsewhere). If `model-api-builder` is too long, the perhaps `api-builder` is good enough too? (I.e. `crud-rest.api-builder.ts`).

Reply:
I like `crud-rest.api-builder.ts` because it's shorter (and the builder is `CrudRestApiBuilder` instead of `CrudRestModelApiBuilder`, so the `Model` part is implied).